crates/currency/src/lib.rs

```diff
@@ -14,23 +15,419 @@ use frame_support::{
+                other,
+                <BalanceOf<T> as CheckedAdd>::checked_add,
+                Error::<T>::ArithmeticOverflow,
+            )
+        }
+
+        pub fn checked_sub(&self, other: &Self) -> Result<Self, DispatchError> {
+            self.checked_fn(
+                other,
+                <BalanceOf<T> as CheckedSub>::checked_sub,
+                Error::<T>::ArithmeticUnderflow,
+            )
+        }
+
+        pub fn saturating_sub(&self, other: &Self) -> Result<Self, DispatchError> {
+            ensure!(self.currency_id == other.currency_id, Error::<T>::InvalidCurrency);
+            self.checked_sub(other)
+                .or_else(|_| Ok(Self::new(0u32.into(), self.currency_id)))
+        }
+
+        pub fn checked_fixed_point_mul(&self, scalar: &UnsignedFixedPoint<T>) -> Result<Self, DispatchError> {
+            let amount = scalar
+                .checked_mul_int(self.amount)
+                .ok_or(Error::<T>::ArithmeticUnderflow)?;
+            Ok(Self {
+                amount,
+                currency_id: self.currency_id,
+            })
+        }
+
+        pub fn checked_fixed_point_mul_rounded_up(
+            &self,
+            scalar: &UnsignedFixedPoint<T>,
+        ) -> Result<Self, DispatchError> {
+            let self_fixed_point =
+                UnsignedFixedPoint::<T>::checked_from_integer(self.amount).ok_or(Error::<T>::TryIntoIntError)?;
+
+            // do the multiplication
+            let product = self_fixed_point
+                .checked_mul(&scalar)
+                .ok_or(Error::<T>::ArithmeticOverflow)?;
+
+            // convert to inner
+            let product_inner = UniqueSaturatedInto::<u128>::unique_saturated_into(product.into_inner());
+
+            // convert to u128 by dividing by a rounded up division by accuracy
+            let accuracy = UniqueSaturatedInto::<u128>::unique_saturated_into(UnsignedFixedPoint::<T>::accuracy());
+            let amount = product_inner
+                .checked_add(accuracy)
+                .ok_or(Error::<T>::ArithmeticOverflow)?
+                .checked_sub(1)
+                .ok_or(Error::<T>::ArithmeticUnderflow)?
+                .checked_div(accuracy)
+                .ok_or(Error::<T>::ArithmeticUnderflow)?
+                .try_into()
+                .map_err(|_| Error::<T>::TryIntoIntError)?;
+
+            Ok(Self {
+                amount,
+                currency_id: self.currency_id,
+            })
+        }
+
+        pub fn checked_div(&self, scalar: &UnsignedFixedPoint<T>) -> Result<Self, DispatchError> {
+            let amount = UnsignedFixedPoint::<T>::checked_from_integer(self.amount)
+                .ok_or(Error::<T>::TryIntoIntError)?
+                .checked_div(&scalar)
+                .ok_or(Error::<T>::ArithmeticOverflow)?
+                .into_inner()
+                .checked_div(&UnsignedFixedPoint::<T>::accuracy())
+                .ok_or(Error::<T>::ArithmeticUnderflow)?;
```

Comment:
Should we extract this logic somewhere? There are a few places we want to divide the inner by the accuracy.

Reply:
Extracted into a `truncate_to_inner` trait implementation